### ISSUE TITLE
Deal with preparation in templates

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2#390]
Currently, the way to do `before_run` preparations so that they are properly supported both in runtime and standalone is to create a new `CodeObject` that is executed in `before_run`. This approach is used for initializing the `SpikeQueue` and for the preparation of `SpatialNeuron` (currently only in the `spatialneuron` branch). This adds a lot of boilerplate code to the main Python classes and also adds to our proliferation of templates. How about having them in the main templates as (optional) blocks? I.e. for every template, you could specify code that is executed before, during and after a run.
I don't see any use case for user-generated code in the before and after run part, currently, but we could always add this option later.
The only disadvantage of this that I'm seeing is that it would make the numpy templates slightly more complicated, since they don't use any blocks at all, currently. On the other hand, this would make them more similar to the other codegen target templates.

**Comment by thesamovar** [comment]
I'm happy with this idea, and I don't think it's a problem to introduce blocks to the numpy templates.

I'm wondering if it would also be possible to have templates in different languages like this? e.g. preparation code in Python but executable code in C++. I think we do this at the moment, right?

**Comment by mstimberg** [comment]
> I'm wondering if it would also be possible to have templates in different languages like this? e.g. preparation code in Python but executable code in C++. I think we do this at the moment, right?

We have a special `python_pre` block that we can use that will call Python code. I'm not sure this is really necessary, though, since all of our runtime targets can also call Python code directly. It'd be slightly involved for weave, if we were calling a function with arguments but currently our only use case is to call `_owner.initialise_queue()`, and this we could also write directly in weave directly (I didn't try it, but something like `_owner.call_method("initialise_queue");` I think.).

**Comment by thesamovar** [comment]
OK I'm happy to go with whatever you think is the nicest solution here.

**Comment by mstimberg** [comment]
I've started looking into this again after doing a few small changes to the templates to make them more consistent and easier to understand. I still think this is a good idea, but I realized that we have two options:
1. We include the preparatory code in the same file as the standard run code but execute it only the first time (with `if` or `try`/`except`).
2. We put the code into a separate file and call it explicitly.

I think I prefer 2 because it seems more direct and would also extend to `after_run` code, but it would add yet another source file and therefore increase compilation time (in particular for Cython). On the other hand, we use this feature very rarely (in most simulations, there will be one additional object/file for each synaptic pathway), so it probably does not have much of an effect in practice.
